controller/instance_manager_controller.go

```diff
@@ -1702,76 +1715,74 @@ func (imc *InstanceManagerController) deleteOrphans(im *longhorn.InstanceManager
-	if autoDelete {
-		return true, nil
-	}
-
-	instanceName := orphan.Spec.Parameters[longhorn.OrphanInstanceName]
-	instanceProc, instanceExist := instanceMap[instanceName]
-	instanceCRScheduledBack, err := isInstanceScheduledOnNode(orphan.Spec.NodeID, instanceName)
-	if err != nil {
-		return false, errors.Wrapf(err, "failed to check CR scheduled node for %v orphan %v", orphan.Spec.Type, instanceName)
-	}
-
-	var instanceState longhorn.InstanceState
-	if instanceExist {
-		instanceState = instanceProc.Status.State
-	} else {
-		instanceState = longhorn.InstanceStateTerminated
+func (imc *InstanceManagerController) deleteOrphan(orphan *longhorn.Orphan) error {
+	imc.logger.Infof("Deleting Orphan %q", orphan.Name)
+	if err := imc.ds.DeleteOrphan(orphan.Name); err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			return nil
+		}
+		return errors.Wrapf(err, "failed to delete orphan %q", orphan.Name)
 	}
+	return nil
+}
 
-	imc.logger.WithFields(logrus.Fields{
-		"instanceExist":           instanceExist,
-		"instanceState":           instanceState,
-		"instanceCRScheduledBack": instanceCRScheduledBack,
-	}).Debugf("Checked deletable for %s orphan instance %s", orphan.Spec.Type, orphan.Name)
-	if !instanceExist || instanceCRScheduledBack {
-		return true, nil
-	}
-	return false, nil
+func (imc *InstanceManagerController) canDeleteOrphanCR(imTerminating, autoDeleteEnabled, instanceExist, instanceCRScheduledBack bool) bool {
+	return imTerminating || autoDeleteEnabled || !instanceExist || instanceCRScheduledBack
 }
 
-func (imc *InstanceManagerController) isEngineScheduledOnCurrentNode(node string, instance string) (bool, error) {
+func (imc *InstanceManagerController) isEngineScheduledInCurrentInstanceManager(instanceManager string, instance string) (bool, error) {
 	existEngine, err := imc.ds.GetEngineRO(instance)
 	switch {
 	case err == nil:
 		// Engine CR still exists - check the ownership
-		return imc.isInstanceScheduledOnCurrentNode(node, &existEngine.Spec.InstanceSpec, &existEngine.Status.InstanceStatus), nil
+		return imc.isInstanceOnInstanceManager(instanceManager, &existEngine.ObjectMeta, &existEngine.Spec.InstanceSpec, &existEngine.Status.InstanceStatus), nil
 	case apierrors.IsNotFound(err):
 		// Engine CR not found - instance is orphaned
 		return false, nil
 	default:
 		// Unexpected error - unable to check if engine instance is orphaned or not
-		return false, err
+		return false, errors.Wrapf(err, "failed to check CR scheduled node for engine orphan %v", instance)
 	}
 }
 
-func (imc *InstanceManagerController) isReplicaScheduledOnCurrentNode(node string, instance string) (bool, error) {
+func (imc *InstanceManagerController) isReplicaScheduledInCurrentInstanceManager(instanceManager string, instance string) (bool, error) {
```

Comment:
Improve reability and fix log. Suggestion also applies for `isEngineScheduledInCurrentInstanceManager`.

```suggestion
	existReplica, err := imc.ds.GetReplicaRO(instance)
	if err != nil {
		if apierrors.IsNotFound(err) {
			// Replica CR not found - instance is orphaned
			return false, nil
		}
		return false, errors.Wrapf(err, "failed to check if instance %q is scheduled on instance manager %q", instance, instanceManager)
	}

	return imc.isInstanceOnInstanceManager(instanceManager, &existReplica.ObjectMeta, &existReplica.Spec.InstanceSpec, &existReplica.Status.InstanceStatus), nil

```